File: importers/figma/src/bindings.ts
/**
 * The phase-2 join (story #167,
 * docs/decisions/token-resolution-phase-split.md): sidecar
 * `{ nodeId, property, variableId }` rows joined against the
 * plugin-exported vartable, each variable's value resolved for the
 * node's mode.
 *
 * The mode is the capture's: a subtree pinned with
 * `explicitVariableModes` resolves that collection's pinned mode; an
 * unpinned subtree resolves the collection's `defaultModeId`. A pinned
 * (non-default) mode qualifies the signal name — `size/gap@dark` — so
 * one variable resolved in two modes yields two signals: the document's
 * mode pins are authored intent, and a runtime producer drives each mode
 * context separately (docs/decisions/binding-table-in-the-document.md).
 *
 * The join knows nothing of binding channels: which property paths are
 * bindable is `dashc`'s Figma-aware half (P5,
 * `crates/dashc/src/figma/bindings.rs`). What this module owns is
 * everything about variables and modes — and every row it cannot join is
 * a named diagnostic, never a silent drop (P4).
 */

import type { ClosureFile, ClosureNode } from "./closure.ts";
import type { ResolvedVarsSidecar } from "./tokens.ts";
import {
  type Vartable,
  type VartableCollection,
  vartableStaleness,
} from "./vartable.ts";

/** One joined row, as the compile request carries it (ABI v2). */
export type JoinedBinding =
  & {
    readonly nodeId: string;
    readonly property: string;
    /** The mode-qualified signal name (`size/gap`, `size/gap@dark`). */
    readonly signal: string;
  }
  & (
    | { readonly resolvedType: "FLOAT"; readonly value: number }
    | {
      readonly resolvedType: "COLOR";
      readonly value: {
        readonly r: number;
        readonly g: number;
        readonly b: number;
        readonly a: number;
      };
    }
  );

/** A named join verdict (P4). */
export interface BindingDiagnostic {
  readonly rule: string;
  readonly severity: "error" | "warning";
  readonly message: string;
  readonly nodeId?: string;
}

/** The join's outcome: the rows that joined, and every named verdict. */
export interface JoinResult {
  readonly bindings: readonly JoinedBinding[];
  readonly diagnostics: readonly BindingDiagnostic[];
}

/**
 * A join the importer refuses to compile past: an error-severity verdict
 * means at least one authored binding cannot be carried faithfully, and
 * emitting a `.dsb` without it would be the silent drop P4 forbids —
 * the same posture as `TokensBlocked`.
 */
export class BindingsBlocked extends Error {
  readonly diagnostics: readonly BindingDiagnostic[];

  constructor(diagnostics: readonly BindingDiagnostic[]) {
    super(
      diagnostics
        .map((d) => `${d.severity}[${d.rule}]: ${d.message}`)
        .join("\n"),
    );
    this.name = "BindingsBlocked";
    this.diagnostics = diagnostics;
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/** The per-collection mode each node resolves in, walked once. */
function effectiveModes(
  file: ClosureFile,
): Map<string, Readonly<Record<string, string>>> {
  const modes = new Map<string, Readonly<Record<string, string>>>();
  const walk = (
    node: ClosureNode,
    inherited: Readonly<Record<string, string>>,
  ): void => {
    const pinned = (node as unknown as Record<string, unknown>)
      .explicitVariableModes;
    const effective = isRecord(pinned)
      ? { ...inherited, ...(pinned as Record<string, string>) }
      : inherited;
    modes.set(node.id, effective);
    for (const child of node.children ?? []) walk(child, effective);
  };
  walk(file.document, {});
  return modes;
}

/**
 * Joins the phase-1 sidecar against the vartable, resolving each
 * variable's value for its node's mode.
 *
 * Named verdicts (P4), all error severity — a binding the designer
 * authored either joins whole or blocks the export:
 *
 * - `figma.vartable.version-mismatch` — the vartable was exported from a
 *   different file version than the sidecar (the staleness guard).
 * - `figma.bindings.unknown-variable` — a sidecar id the vartable does
 *   not carry (the join is not total).
 * - `figma.bindings.no-mode-value` — the resolved mode has no value for
 *   the variable.
 * - `figma.bindings.alias-value` — the mode's value is itself a variable
 *   alias; alias chains are not resolved this slice.
 * - `figma.bindings.value-shape` — a FLOAT value that is not a number,
 *   or a COLOR value that is not an `{r,g,b,a}` object.
 * - `figma.bindings.ambiguous-signal` — two variables (or one variable
 *   in two collections' name spaces) yield one signal name.
 *
 * One warning: `figma.bindings.unsupported-type` — a STRING or BOOLEAN
 * variable. Text, variant, and visibility bindings are later slices; the
 * site is named and the resolved literal still ships.
 */
export function joinBindings(
  sidecar: ResolvedVarsSidecar,
  vartable: Vartable,
  file: ClosureFile,
): JoinResult {
  const diagnostics: BindingDiagnostic[] = [];
  const bindings: JoinedBinding[] = [];

  const stale = vartableStaleness(vartable, sidecar.version);
  if (stale !== null) {
    return { bindings, diagnostics: [stale] };
  }

  const modes = effectiveModes(file);
  /** signal name → the variable id + mode that first claimed it. */
  const claimed = new Map<string, string>();

  const fail = (rule: string, message: string, nodeId: string): void => {
    diagnostics.push({ rule, severity: "error", message, nodeId });
  };

  for (const row of sidecar.bindings) {
    const variable = vartable.variables[row.variableId];
    if (variable === undefined) {
      fail(
        "figma.bindings.unknown-variable",
        `${row.property}: variable ${row.variableId} has no vartable entry — ` +
          `the join is not total; re-run the annotator's token export`,
        row.nodeId,
      );
      continue;
    }
    if (
      variable.resolvedType !== "FLOAT" && variable.resolvedType !== "COLOR"
    ) {
      diagnostics.push({
        rule: "figma.bindings.unsupported-type",
        severity: "warning",
        message: `${row.property}: variable "${variable.name}" is ` +
          `${variable.resolvedType}; only FLOAT and COLOR bindings are ` +
          `carried this slice — the resolved literal still ships`,
        nodeId: row.nodeId,
      });
      continue;
    }

    const collection: VartableCollection | undefined =
      vartable.collections[variable.variableCollectionId];
    if (collection === undefined) {
      fail(
        "figma.bindings.unknown-variable",
        `${row.property}: variable "${variable.name}" references collection ` +
          `${variable.variableCollectionId}, which the vartable does not carry`,
        row.nodeId,
      );
      continue;
    }

    const pinned = modes.get(row.nodeId)?.[collection.id];
    const modeId = pinned ?? collection.defaultModeId;
    const value = variable.valuesByMode[modeId];
    if (value === undefined) {
      fail(
        "figma.bindings.no-mode-value",
        `${row.property}: variable "${variable.name}" has no value for mode ` +
          `${modeId} of collection "${collection.name}"`,
        row.nodeId,
      );
      continue;
    }
    if (isRecord(value) && value.type === "VARIABLE_ALIAS") {
      fail(
        "figma.bindings.alias-value",
        `${row.property}: variable "${variable.name}" resolves to another ` +
          `variable in mode ${modeId}; alias chains are not resolved this slice`,
        row.nodeId,
      );
      continue;
    }

    // A non-default resolved mode qualifies the signal name, so one
    // variable in two mode contexts is two signals.
    const modeName = collection.modes.find((m) => m.modeId === modeId)?.name ??
      modeId;
    const signal = modeId === collection.defaultModeId
      ? variable.name
      : `${variable.name}@${modeName}`;

    const claim = `${row.variableId}@${modeId}`;
    const earlier = claimed.get(signal);
    if (earlier !== undefined && earlier !== claim) {
      fail(
        "figma.bindings.ambiguous-signal",
        `signal name "${signal}" resolves from two different variables or ` +
          `modes (${earlier} and ${claim}); rename one variable`,
        row.nodeId,
      );
      continue;
    }
    claimed.set(signal, claim);

    if (variable.resolvedType === "FLOAT") {
      if (typeof value !== "number") {
        fail(
          "figma.bindings.value-shape",
          `${row.property}: FLOAT variable "${variable.name}" carries a ` +
            `non-number value in mode ${modeId}`,
          row.nodeId,
        );
        continue;
      }
      bindings.push({
        nodeId: row.nodeId,
        property: row.property,
        signal,
        resolvedType: "FLOAT",
        value,
      });
    } else {
      if (
        !isRecord(value) ||
        typeof value.r !== "number" || typeof value.g !== "number" ||
        typeof value.b !== "number" || typeof value.a !== "number"
      ) {
        fail(
          "figma.bindings.value-shape",
          `${row.property}: COLOR variable "${variable.name}" carries a ` +
            `non-color value in mode ${modeId}`,
          row.nodeId,
        );
        continue;
      }
      bindings.push({
        nodeId: row.nodeId,
        property: row.property,
        signal,
        resolvedType: "COLOR",
        value: { r: value.r, g: value.g, b: value.b, a: value.a },
      });
    }
  }

  return { bindings, diagnostics };
}

File: crates/dashc/src/figma/bindings.rs
//! Applying the importer's joined variable-binding rows to a lowered
//! document (story #167; `docs/decisions/token-resolution-phase-split.md`,
//! `docs/decisions/binding-table-in-the-document.md`).
//!
//! The importer joins the phase-1 sidecar (`{nodeId, property,
//! variableId}`) against the plugin-exported vartable and resolves each
//! variable's value for the node's mode; what crosses the ABI is one
//! [`BoundVariable`] per site: the Figma node id, the Figma property
//! path, the mode-qualified signal name, and the resolved value. This
//! module is the Figma-aware half of the split (P5): it knows which
//! property paths map onto which binding channels; the importer knows
//! nothing of channels, and this module knows nothing of modes.
//!
//! Every unsupported site is a named diagnostic, never a silent drop
//! (P4) — and never an error: the lowering already emitted the site's
//! resolved literal, so the picture is right; only the *live* binding is
//! not carried yet.

use std::collections::{BTreeMap, HashMap};

use dashpaint::PaintKind;
use dashscene_validator::{Diagnostic, Location, NodePath, Severity};

use crate::document::{Binding, BindingChannel, BindingTransform, Document, SignalDecl};

/// The diagnostic rules this producer assembles for binding rows.
pub mod rule {
    /// The row names a Figma node the lowering emitted no document node
    /// for — a component definition (resolves but does not paint), or a
    /// node id the capture does not carry. The binding has no target, so
    /// it is named and not emitted; nothing rendered is affected.
    pub const UNLOWERED_NODE: &str = "figma.bindings.unlowered-node";
    /// The row's property path has no binding channel in the vocabulary
    /// yet (a corner radius, a stroke color, a gradient stop, an effect,
    /// an opacity), or its value type does not fit the path. The resolved
    /// literal ships; the live binding does not — named, never silent.
    pub const UNSUPPORTED_PROPERTY: &str = "figma.bindings.unsupported-property";
    /// Two rows declare one signal name with two different initial
    /// values. The join qualifies a signal's name by its resolved mode,
    /// so one name always resolves to one value — a conflict means the
    /// rows are inconsistent, and emitting either value would silently
    /// pick one (P4).
    pub const SIGNAL_CONFLICT: &str = "figma.bindings.signal-conflict";
}

/// One joined variable-binding row, as decoded from the compile request.
#[derive(Debug, Clone, PartialEq)]
pub struct BoundVariable {
    /// The Figma node id the sidecar recorded (`"1:8"`).
    pub node_id: String,
    /// The sidecar property path (`"itemSpacing"`, `"fills[0].color"`).
    pub property: String,
    /// The mode-qualified signal name the join produced
    /// (`"size/gap"`, `"size/gap@dark"`). A color site appends the
    /// channel suffix here (`.r`/`.g`/`.b`/`.a`).
    pub signal: String,
    /// The variable's value, resolved for the node's mode.
    pub value: BoundValue,
}

/// A joined row's resolved value. `FLOAT` and `COLOR` are the two
/// variable types with a channel vocabulary this slice; the importer
/// names `STRING`/`BOOLEAN` rows itself and does not send them.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum BoundValue {
    Float(f32),
    Color { r: f32, g: f32, b: f32, a: f32 },
}

/// Where one lowered node landed, and what the binding rows need to know
/// about how it lowered.
pub(super) struct LoweredNode {
    /// The node's document index.
    pub(super) index: u32,
    /// The node's diagnostic name path.
    pub(super) path: String,
    /// The visible fill's paint `opacity` (1.0 when absent). The lowering
    /// folds it into the shipped literal's alpha (`color_of`:
    /// `color.a * opacity`), so a FillA binding must capture the same
    /// multiply as its transform — otherwise the seeded scene would jump
    /// to the raw variable alpha.
    pub(super) fill_opacity: f32,
}

/// Figma node id → how it lowered, for every node the walk emitted.
pub(super) type IndexOfId = BTreeMap<String, LoweredNode>;

/// Applies the joined rows to the lowered document, appending
/// `Document.signals` and `Document.bindings`, and returns the
/// diagnostics the rows earned. Rows apply in the order given (the
/// sidecar's document order), and signals intern in first-use order, so
/// the same input yields the same tables (R7).
pub(super) fn apply(
    doc: &mut Document,
    rows: &[BoundVariable],
    index_of_id: &IndexOfId,
) -> Vec<Diagnostic> {
    let mut diagnostics = Vec::new();
    let mut signal_of: HashMap<String, u32> = HashMap::new();

    for row in rows {
        let Some(lowered) = index_of_id.get(&row.node_id) else {
            diagnostics.push(Diagnostic {
                rule: rule::UNLOWERED_NODE,
                severity: Severity::Warning,
                at: Location::Node(NodePath::new(0, format!("<figma:{}>", row.node_id))),
                message: format!(
                    "binding of \"{}\" targets Figma node {}, which lowered to no document node \
                     (a definition, or a node outside the export); the binding is not carried",
                    row.signal, row.node_id
                ),
            });
            continue;
        };
        let (node, path) = (&lowered.index, &lowered.path);
        let at = || Location::Node(NodePath::new(*node, path.clone()));
        let unsupported = |what: &str| Diagnostic {
            rule: rule::UNSUPPORTED_PROPERTY,
            severity: Severity::Warning,
            at: at(),
            message: format!(
                "{what}; the resolved literal ships in the document, but the live binding of \
                 \"{}\" is not carried",
                row.signal
            ),
        };

        match (&row.value, classify_property(&row.property)) {
            (BoundValue::Float(v), PropertySite::ItemSpacing) => {
                if doc.nodes[*node as usize].container.is_none() {
                    diagnostics.push(unsupported(&format!(
                        "itemSpacing is bound on {path}, but the node lowered without an \
                         auto-layout container"
                    )));
                    continue;
                }
                match intern_signal(doc, &mut signal_of, &row.signal, *v) {
                    Ok(signal) => doc.bindings.push(Binding {
                        signal,
                        node: *node,
                        channel: BindingChannel::Gap,
                        transform: BindingTransform::Identity,
                    }),
                    Err(diagnostic) => diagnostics.push(diagnostic_at(diagnostic, at())),
                }
            }
            (BoundValue::Color { r, g, b, a }, PropertySite::FillColor) => {
                let solid = matches!(
                    doc.nodes[*node as usize].paint,
                    Some(ref paint) if matches!(paint.entry.fill, Some(PaintKind::Solid { .. }))
                );
                if !solid {
                    diagnostics.push(unsupported(&format!(
                        "a fill color is bound on {path}, but the node lowered without a solid \
                         fill (a text node's fill lives in its text style)"
                    )));
                    continue;
                }
                // The lowering folds the paint's opacity into the shipped
                // literal's alpha (`color_of`), so the alpha channel's
                // transform captures the same multiply over the raw
                // variable alpha — the transform-applied initial equals
                // the literal, and a producer pushing a new variable
                // value stays folded. The other components are unfolded.
                let alpha_transform = if lowered.fill_opacity == 1.0 {
                    BindingTransform::Identity
                } else {
                    BindingTransform::Scale(lowered.fill_opacity)
                };
                let components: [(&str, BindingChannel, f32, BindingTransform); 4] = [
                    (".r", BindingChannel::FillR, *r, BindingTransform::Identity),
                    (".g", BindingChannel::FillG, *g, BindingTransform::Identity),
                    (".b", BindingChannel::FillB, *b, BindingTransform::Identity),
                    (".a", BindingChannel::FillA, *a, alpha_transform),
                ];
                for (suffix, channel, component, transform) in components {
                    let name = format!("{}{suffix}", row.signal);
                    match intern_signal(doc, &mut signal_of, &name, component) {
                        Ok(signal) => doc.bindings.push(Binding {
                            signal,
                            node: *node,
                            channel,
                            transform,
                        }),
                        Err(diagnostic) => diagnostics.push(diagnostic_at(diagnostic, at())),
                    }
                }
            }
            (BoundValue::Float(v), PropertySite::Opacity) => {
                // Node/group opacity binds on any node — no container or
                // fill precondition, unlike gap and fill color (story #44,
                // debt #253). The natural landing for a Figma opacity
                // variable now that the document carries `Node.opacity`.
                match intern_signal(doc, &mut signal_of, &row.signal, *v) {
                    Ok(signal) => doc.bindings.push(Binding {
                        signal,
                        node: *node,
                        channel: BindingChannel::Opacity,
                        transform: BindingTransform::Identity,
                    }),
                    Err(diagnostic) => diagnostics.push(diagnostic_at(diagnostic, at())),
                }
            }
            (value, site) => {
                let kind = match value {
                    BoundValue::Float(_) => "FLOAT",
                    BoundValue::Color { .. } => "COLOR",
                };
                let what = match site {
                    PropertySite::ItemSpacing | PropertySite::FillColor | PropertySite::Opacity => {
                        format!(
                            "a {kind} variable is bound to {}, which takes the other value type",
                            row.property
                        )
                    }
                    PropertySite::Other => format!(
                        "{} has no binding channel in the vocabulary yet",
                        row.property
                    ),
                };
                diagnostics.push(unsupported(&what));
            }
        }
    }

    diagnostics
}

/// A partially-built signal-conflict diagnostic, completed with the
/// node location by the caller.
struct SignalConflict {
    message: String,
}

fn diagnostic_at(conflict: SignalConflict, at: Location) -> Diagnostic {
    Diagnostic {
        rule: rule::SIGNAL_CONFLICT,
        severity: Severity::Error,
        at,
        message: conflict.message,
    }
}

/// Interns one signal declaration by name, first-use order. A repeated
/// name must repeat its initial value bit-for-bit: the join qualifies
/// names by mode, so one name is one value, and a mismatch is an
/// inconsistent row set.
fn intern_signal(
    doc: &mut Document,
    signal_of: &mut HashMap<String, u32>,
    name: &str,
    initial: f32,
) -> Result<u32, SignalConflict> {
    if let Some(&index) = signal_of.get(name) {
        let declared = doc.signals[index as usize].initial;
        if declared.to_bits() != initial.to_bits() {
            return Err(SignalConflict {
                message: format!(
                    "signal \"{name}\" is declared with initial {declared} and bound again with \
                     {initial}; one mode-qualified name resolves to one value, so the joined \
                     rows are inconsistent"
                ),
            });
        }
        return Ok(index);
    }
    let index = u32::try_from(doc.signals.len()).expect("document exceeds u32::MAX signals");
    doc.signals.push(SignalDecl {
        name: name.to_owned(),
        initial,
    });
    signal_of.insert(name.to_owned(), index);
    Ok(index)
}

/// The property paths with a channel mapping this slice.
enum PropertySite {
    /// `itemSpacing` — the auto-layout gap.
    ItemSpacing,
    /// `fills[i].color` — a paint's own solid color. Hidden paints are
    /// not in the sidecar and stacked visible fills refuse the compile,
    /// so any index here is the single lowered fill.
    FillColor,
    /// `opacity` — the node/group alpha (story #44, debt #253).
    Opacity,
    Other,
}

fn classify_property(property: &str) -> PropertySite {
    if property == "itemSpacing" {
        return PropertySite::ItemSpacing;
    }
    if property == "opacity" {
        return PropertySite::Opacity;
    }
    if let Some(rest) = property.strip_prefix("fills[")
        && let Some(index) = rest.strip_suffix("].color")
        && index.chars().all(|c| c.is_ascii_digit())
        && !index.is_empty()
    {
        return PropertySite::FillColor;
    }
    PropertySite::Other
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn property_paths_classify_by_shape() {
        assert!(matches!(
            classify_property("itemSpacing"),
            PropertySite::ItemSpacing
        ));
        assert!(matches!(
            classify_property("fills[0].color"),
            PropertySite::FillColor
        ));
        assert!(matches!(
            classify_property("fills[12].color"),
            PropertySite::FillColor
        ));
        assert!(matches!(
            classify_property("opacity"),
            PropertySite::Opacity
        ));
        for other in [
            "strokes[0].color",
            "fills[0].gradientStops[2].color",
            "effects[0].color",
            "rectangleCornerRadii.RECTANGLE_TOP_LEFT_CORNER_RADIUS",
            "fills[].color",
            "fills[x].color",
        ] {
            assert!(
                matches!(classify_property(other), PropertySite::Other),
                "{other} must not classify as a supported site"
            );
        }
    }
}
